src/equations_set_routines.F90

```diff
@@ -1792,22 +1795,59 @@ SUBROUTINE EQUATIONS_SET_BACKSUBSTITUTE(equationsSet,BOUNDARY_CONDITIONS,err,err
                                               & ROW_INDICES,COLUMN_INDICES,err,error,*999)
                                             !Loop over the non-ghosted rows in the equations set
                                             DO equations_row_number=1,vectorMapping%numberOfRows
+                                              ! Same as other case (BLOCK)  
                                               RHS_VALUE=0.0_DP
                                               rhs_variable_dof=rhsMapping%equationsRowToRHSDOFMap(equations_row_number)
                                               rhs_global_dof=RHS_DOMAIN_MAPPING%LOCAL_TO_GLOBAL_MAP(rhs_variable_dof)
                                               rhs_boundary_condition=RHS_BOUNDARY_CONDITIONS%DOF_TYPES(rhs_global_dof)
                                               SELECT CASE(rhs_boundary_condition)
                                               CASE(BOUNDARY_CONDITION_DOF_FREE)
                                                 !Back substitute
-                                                !Loop over the local columns of the equations matrix
+                                                !Loop over the global columns of the equations matrix in the selected row
                                                 DO equations_column_idx=ROW_INDICES(equations_row_number), &
                                                   ROW_INDICES(equations_row_number+1)-1
+                                                  ! Get the GLOBAL column index
                                                   equations_column_number=COLUMN_INDICES(equations_column_idx)
-                                                  variable_dof=equations_column_idx-ROW_INDICES(equations_row_number)+1
-                                                  MATRIX_VALUE=equationsMatrixData(equations_column_idx)
-                                                  DEPENDENT_VALUE=DEPENDENT_PARAMETERS(variable_dof)
-                                                  RHS_VALUE=RHS_VALUE+MATRIX_VALUE*DEPENDENT_VALUE
+                                                  ! Get the LOCAL column index for all domains
+                                                  DO domain_idx=1,COLUMN_DOMAIN_MAPPING%GLOBAL_TO_LOCAL_MAP &
+                                                    & (equations_column_number)%NUMBER_OF_DOMAINS
+                                                    ! Select the current domain (=rank)
+                                                    IF (COLUMN_DOMAIN_MAPPING%GLOBAL_TO_LOCAL_MAP  &
```

Comment:
What happens if GLOBAL_TO_LOCAL_MAP condition is not satisfied?  We will leave RHS_VALUE at 0.0 then

Reply:
Yes, we only need to integrate the dof if it belongs to the current rank. If it doesn't, it will be eventually integrated in its own rank.